src/Analyzer/Passes/AnyFunctionPass.cpp

```diff
@@ -0,0 +1,139 @@
+#include "AnyFunctionPass.h"
+
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <AggregateFunctions/IAggregateFunction.h>
+
+#include <Analyzer/InDepthQueryTreeVisitor.h>
+#include <Analyzer/FunctionNode.h>
+#include <Analyzer/LambdaNode.h>
+#include <Analyzer/ConstantNode.h>
+#include <Analyzer/ArrayJoinNode.h>
+
+namespace DB
+{
+
+namespace
+{
+
+class AnyFunctionVisitor : public InDepthQueryTreeVisitor<AnyFunctionVisitor>
+{
+private:
+    bool canRewrite(const FunctionNode * function_node)
+    {
+        for (const auto & argument : function_node->getArguments().getNodes())
+        {
+            /// arrayJoin() is special and should not be optimized (think about
+            /// it as a an aggregate function), otherwise wrong result will be
+            /// produced:
+            ///     SELECT *, any(arrayJoin([[], []])) FROM numbers(1) GROUP BY number
+            ///     ┌─number─┬─arrayJoin(array(array(), array()))─┐
+            ///     │      0 │ []                                 │
+            ///     │      0 │ []                                 │
+            ///     └────────┴────────────────────────────────────┘
+            /// While should be:
+            ///     ┌─number─┬─any(arrayJoin(array(array(), array())))─┐
+            ///     │      0 │ []                                      │
+            ///     └────────┴─────────────────────────────────────────┘
+            if (argument->as<LambdaNode>())
+                return false;
+
+            if (argument->as<ArrayJoinNode>())
```

Comment:
This is always false. `ArrayJoinNode` relates to `ARRAY JOIN` syntax, not `arrayJoin` function.

Reply:
Also, the comment, in the beginning, sounds quite strange because this "optimization" changes the behavior of the query..